core/modules/entity/entity.module

```diff
@@ -739,15 +739,79 @@ function entity_view_mode_prepare($entity_type, $entities, $view_mode, $langcode
   return $entities_by_view_mode;
 }
 
+/**
+ * Implements hook_preprocess().
+ *
+ * Add template suggestions for an entity based on view modes if they do not
+ * already exist.
+ *
+ * The eventual order of the entity's theme suggestions are:
+ * - entity-type__bundle (lowest priority = first in the array)
+ * - entity-type__bundle__view-mode
+ * - entity-type__id
+ * - entity-type__id__view-mode (highest priority = last in the array)
+ */
+function entity_preprocess(&$variables, $hook) {
```

Comment:
Too bad preprocess function this gets called for every single `hook_theme()` implementation that uses a template (e.g. blocks, layouts, etc). 

I'm wondering if there is a better way to handle this. For example, Backdrop already provides `entity` as a theme hook (in `entity_theme()`) that AFAIK never gets used at all. 

(UNTESTED) We could modify that by using the key `pattern` to provide a matching pattern for each core entity, as in 


```php
  $types = array();
  foreach (entity_get_info() as $type => $info) {
    if (isset($info['controller class']) && in_array('EntityPlusControllerInterface', class_implements($info['controller class']))) {
      $types[] = $type;
    }
  }
  $pattern = '(' . implode('|', $types) . ')(\.|__)';
```
And then provide that in the return of `entity_theme()` as:

```php
return array(
    'entity' => array(
      'render element' => 'elements',
      'template' => 'templates/entity',
      'file' => 'entity.theme.inc',
      'pattern' => $pattern,
    ),
```

This MAY (untested) allow us to use `entity_preprocess_entity()` instead of `entity_preprocess()`, which would save some calls.